{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import json\n",
    "import warnings\n",
    "import operator\n",
    "\n",
    "import h5py\n",
    "from keras.models import model_from_json\n",
    "from keras import backend as K\n",
    "from keras.utils import get_custom_objects\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "size_title = 18\n",
    "size_label = 14\n",
    "n_pred = 2\n",
    "\n",
    "\n",
    "def read_file(file_path):\n",
    "    with open(file_path, 'r') as data_file:\n",
    "        data = json.loads(data_file.read())\n",
    "    return data\n",
    "\n",
    "def create_model(model_path):\n",
    "    \n",
    "    reverse_dictionary = dict((str(v), k) for k, v in dictionary.items())\n",
    "    model_weights = list()\n",
    "    weight_ctr = 0\n",
    "    while True:\n",
    "        try:\n",
    "            d_key = \"weight_\" + str(weight_ctr)\n",
    "            weights = trained_model.get(d_key).value\n",
    "            model_weights.append(weights)\n",
    "            weight_ctr += 1\n",
    "        except Exception as exception:\n",
    "            break\n",
    "    # set the model weights\n",
    "    loaded_model.set_weights(model_weights)\n",
    "    return loaded_model, dictionary, reverse_dictionary, compatibile_tools\n",
    "\n",
    "\n",
    "def verify_model(model, tool_sequence, labels, dictionary, reverse_dictionary, compatible_tools, class_weights, topk=20, max_seq_len=25):\n",
    "    tl_seq = tool_sequence.split(\",\")\n",
    "    last_tool_name = reverse_dictionary[str(tl_seq[-1])]\n",
    "    last_compatible_tools = compatible_tools[last_tool_name]\n",
    "    sample = np.zeros(max_seq_len)\n",
    "    for idx, tool_id in enumerate(tl_seq):\n",
    "        sample[idx] = int(tool_id)\n",
    "    sample_reshaped = np.reshape(sample, (1, max_seq_len))\n",
    "\n",
    "    tool_sequence_names = [reverse_dictionary[str(tool_pos)] for tool_pos in tool_sequence.split(\",\")]\n",
    "    \n",
    "    # predict next tools for a test path\n",
    "    prediction = model.predict(sample_reshaped, verbose=0)\n",
    "    \n",
    "    weight_val = list(class_weights.values())\n",
    "    weight_val = np.reshape(weight_val, (len(weight_val),))\n",
    "    \n",
    "    prediction = np.reshape(prediction, (prediction.shape[1],))\n",
    "    \n",
    "    #prediction = prediction * weight_val\n",
    "    \n",
    "    prediction = prediction / float(np.max(prediction))\n",
    "    \n",
    "    prediction_pos = np.argsort(prediction, axis=-1)\n",
    "\n",
    "    # get topk prediction\n",
    "    topk_prediction_pos = prediction_pos[-topk:]\n",
    "    topk_prediction_val = [int(prediction[pos] * 100) for pos in topk_prediction_pos]\n",
    "    \n",
    "    # read tool names using reverse dictionary\n",
    "    pred_tool_ids = [reverse_dictionary[str(tool_pos)] for tool_pos in topk_prediction_pos]\n",
    "    actual_next_tool_ids = list(set(pred_tool_ids).intersection(set(last_compatible_tools.split(\",\"))))\n",
    "\n",
    "    pred_tool_ids_sorted = dict()\n",
    "    for (tool_pos, tool_pred_val) in zip(topk_prediction_pos, topk_prediction_val):\n",
    "        tool_name = reverse_dictionary[str(tool_pos)]\n",
    "        if tool_name in actual_next_tool_ids and tool_name not in last_tool_name:\n",
    "            pred_tool_ids_sorted[tool_name] = tool_pred_val\n",
    "    pred_tool_ids_sorted = dict(sorted(pred_tool_ids_sorted.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    \n",
    "    cls_wt = dict()\n",
    "    usg_wt = dict()\n",
    "    inv_wt = dict()\n",
    "    ids_tools = dict()\n",
    "    keys = list(pred_tool_ids_sorted.keys())\n",
    "    for k in keys:\n",
    "        try:\n",
    "            cls_wt[k] = np.round(class_weights[str(data_dict[k])], 2)\n",
    "            usg_wt[k] = np.round(usage_weights[k], 2)\n",
    "            inv_wt[k] = np.round(inverted_weights[str(data_dict[k])], 2)\n",
    "        except:\n",
    "            continue\n",
    "    print(\"Predicted tools: \\n\")\n",
    "    print(pred_tool_ids_sorted)\n",
    "    print()\n",
    "    print(\"Class weights: \\n\")\n",
    "    cls_wt = dict(sorted(cls_wt.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    print(cls_wt)\n",
    "    print()\n",
    "    print(\"Usage weights: \\n\")\n",
    "    usg_wt = dict(sorted(usg_wt.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    print(usg_wt)\n",
    "    print()\n",
    "    total_usage_wt = np.mean(list(usg_wt.values()))\n",
    "    print(\"Mean usage wt: %0.4f\" % (total_usage_wt))\n",
    "    print()\n",
    "    print(\"Inverted weights: \\n\")\n",
    "    inv_wt = dict(sorted(inv_wt.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    print(inv_wt)\n",
    "    for key in pred_tool_ids_sorted:\n",
    "        ids_tools[key] = dictionary[key]\n",
    "    print()\n",
    "    print(\"Tool ids\")\n",
    "    print(ids_tools)\n",
    "    print(\"======================================\")\n",
    "    return cls_wt, usg_wt, inv_wt, pred_tool_ids_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': 'Add_a_column1', '2': 'bedtools_genomecoveragebed_bedgraph', '3': 'hmmer_hmmsearch', '4': 'cshl_awk_tool', '5': 'gff_to_sequence', '6': 'rbc_mirdeep2_quantifier', '7': 'cshl_fastx_barcode_splitter', '8': 'smf_utils_extract-boxed-sequences', '9': 'mimodd_reheader', '10': 'graphclust_glob_report_no_align', '11': 'deeptools_plot_heatmap', '12': 'deeptools_bamCorrelate', '13': 'bgchem_fragment_merger', '14': 'sqlite_to_tabular', '15': 'Determine_BC', '16': 'gatk2_haplotype_caller', '17': 'viennarna_rnaalifold', '18': 'bedtools_multicovtbed', '19': 'humann2_genefamilies_genus_level', '20': 'ctb_compound_convert', '21': 'idpquery', '22': 'mass_spectrometry_imaging_combine', '23': 'sm_mapper', '24': 'ctb_chemfp_nxn_clustering', '25': 'mothur_count_groups', '26': 'bedtools_sortbed', '27': 'minfi_methcpg', '28': 'EMBOSS: fuzztran39', '29': 'XY_Plot_1', '30': 'pureclip', '31': 'wsdl_hmdb', '32': 'flash', '33': 'fastq_dump', '34': 'sklearn_svm_classifier', '35': 'barchart_gnuplot', '36': 'CONVERTER_bed_to_bgzip_0', '37': 'mimodd_varextract', '38': 'trinity_run_de_analysis', '39': 'mothur_dist_shared', '40': 'bedtools_slopbed', '41': 'deeptools_bamCoverage', '42': 'bg_protein_properties', '43': '__FLATTEN__', '44': 'rarefaction', '45': 'deeptools_compute_matrix_operations', '46': 'bwa_mem', '47': 'hicexplorer_hicpca', '48': 'EMBOSS: getorf42', '49': 'stacks_genotypes', '50': 'MassCalculator', '51': 'tophat_fusion_post', '52': 'r_correlation_matrix', '53': 'msnbase_readmsdata', '54': 'PicardASMetrics', '55': 'ctb_sdf2fps', '56': 'qiime_summarize_taxa_through_plots', '57': 'bcftools_norm', '58': 'cd_hit_protein', '59': 'mothur_pre_cluster', '60': 'fastq_paired_end_joiner', '61': 'rsem_prepare_reference', '62': 'mtbls520_18_phylogeny', '63': 'ctb_remSmall', '64': 'preproc', '65': 'msconvert_win', '66': 'charts', '67': 'selectsequencesfrommsa', '68': 'racon', '69': 'samtools_mpileup', '70': 'cshl_multijoin', '71': 'MapAlignerPoseClustering', '72': 'picard_FilterSamReads', '73': 'ctb_alignit_create_db', '74': 'term_id_vs_term_name', '75': 'PeptideIndexer', '76': 'vcfcheck', '77': 'tabular_to_fastq', '78': 'bam2wig', '79': 'bg_uniq', '80': 'glimmer_not-knowlegde-based', '81': 'signalp3', '82': 'methtools_tiling', '83': 'qiime_count_seqs', '84': 'lfq_protein_quant', '85': 'hmmer_hmmbuild', '86': 'abims_xcms_summary', '87': 'fasplit', '88': 'cshl_fasta_nucleotides_changer', '89': 'umi_tools_whitelist', '90': '__MERGE_COLLECTION__', '91': 'EMBOSS: shuffleseq87', '92': 'trim_reads', '93': 'mass_spectrometry_imaging_mzplots', '94': 'dunovo', '95': 'blastxml_to_tabular', '96': 'qiime_core_diversity', '97': 'GeMoMa_Annotation_Filter', '98': 'deeptools_bam_compare', '99': 'ncbi_tblastn_wrapper', '100': 'abims_xcms_group', '101': 'stringtie', '102': 'ChooseTag', '103': 'mycrobiota-qc-report', '104': 'IsobaricAnalyzer', '105': 'tp_uniq_tool', '106': 'bcftools_view', '107': 'ggplot2_point', '108': 'mothur_seq_error', '109': 'OpenSwathAssayGenerator', '110': 'meme_meme', '111': 'FROGS_affiliation_OTU', '112': 'mothur_remove_groups', '113': 'mtbls520_08e_species_features', '114': 'MzTabExporter', '115': 'CONVERTER_fasta_to_fai', '116': 'Extractor', '117': 'w4mcorcov', '118': 'tab2fasta', '119': 'cshl_princeton_fastx_barcode_splitter', '120': 'CONVERTER_gff_to_interval_index_0', '121': 'ctb_ob_genProp', '122': 'scanpy_run_pca', '123': 'mothur_make_shared', '124': 'fasta_filter_by_length', '125': 'tp_sort_header_tool', '126': 'cat1', '127': 'Remove_ending', '128': 'vcfbedintersect', '129': 'bed_to_protein_map', '130': 'xcms_plot_chromatogram', '131': 'ip_convertimage', '132': 'trinity_filter_low_expr_transcripts', '133': 'rseqc_junction_saturation', '134': 'qiime_make_otu_heatmap', '135': 'trim_galore', '136': 'map-msms2camera', '137': 'gops_basecoverage_1', '138': 'snippy', '139': 'SequenceCoverageCalculator', '140': 'get_subontology_from', '141': 'vsearch_chimera_detection', '142': 'datamash_transpose', '143': 'deeptools_plot_coverage', '144': 'bcftools_call', '145': 'fasta-stats', '146': 'mcClust', '147': 'cshl_fastq_to_fasta', '148': 'mimodd_convert', '149': 'diffbind-embl', '150': 'shyft_longitudes', '151': 'keras_model_config', '152': 'infernal_cmsearch', '153': 'MultiplexResolver', '154': 'fasta2tab', '155': 'gffcompare', '156': 'computeMatrix', '157': 'jcvi_gff_stats', '158': 'TrimByLength', '159': 'dexseq_count', '160': 'scanpy_run_tsne', '161': 'mothur_get_seqs', '162': 'scatterplot_rpy', '163': 'rna_star', '164': 'camera-prepareoutput', '165': 'meme_chip', '166': 'gatk2_variant_recalibrator', '167': 'comp1', '168': 'make_families', '169': 'rgweblogo3', '170': 'hicexplorer_hicfindtads', '171': 'metaquantome_filter', '172': 'stoceps_maketablecarrer', '173': 'mtbls520_08b_species_unique', '174': 'datamash_ops', '175': 'stacks_populations', '176': 'deeptools_correctGCBias', '177': 'ctb_simsearch', '178': 'mothur_pairwise_seqs', '179': 'gemini_windower', '180': 'mothur_summary_tax', '181': 'seqtk_subseq', '182': 'FROGS_affiliations_stat', '183': 'clustalw', '184': 'mtbls520_08c_species_variability', '185': 'mothur_cluster', '186': 'bedtools_intersectbed_bam', '187': 'megahit', '188': 'minfi_dmp', '189': 'tables_arithmetic_operations', '190': 'regionalgam_flight_curve', '191': 'ctb_shapeit', '192': 'abricate', '193': 'gatk2_realigner_target_creator', '194': 'dexseq', '195': 'cufflinks_prok', '196': 'Digestor', '197': 'bam_to_sam', '198': 't_coffee', '199': 'fastq_paired_end_interlacer', '200': 'fasta_merge_files_and_filter_unique_sequences', '201': 'picard_MarkDuplicatesWithMateCigar', '202': 'GeneModelMapper', '203': 'prepare_receptor', '204': 'gafa', '205': '16Sclassifier', '206': 'DecoyDatabase', '207': 'deeptools_bigwig_compare', '208': 'deeptools_plot_profile', '209': 'bedtools_multiintersectbed', '210': 'gmx_solvate', '211': 'cardinal_preprocessing', '212': 'miniasm', '213': 'w4mclstrpeakpics', '214': 'newick_display', '215': 'fastq_quality_trimmer', '216': 'Multivariate', '217': 'xcms-find-peaks', '218': 'augustus', '219': 'cd_hit_dup', '220': '16Saligner', '221': 'pynast', '222': 'mothur_get_oturep', '223': 'mz_to_sqlite', '224': 'join1', '225': 'vcftools_isec', '226': 'smf_utils_filter-by-energy', '227': 'bg_sortmerna', '228': 'gatk2_variant_filtration', '229': 'qiime_assign_taxonomy', '230': 'gops_subtract_1', '231': 'secretbt2test', '232': 'gops_coverage_1', '233': 'unipept', '234': 'seurat_export_cellbrowser', '235': 'graphicsmagick_image_convert', '236': 'table_compute', '237': 'chipseeker', '238': 'trinity_contig_exn50_statistic', '239': 'bedtools_mergebed', '240': 'qiime_validate_mapping_file', '241': 'mothur_remove_seqs', '242': 'taxonomy_krona_chart', '243': 'mtbls520_07_species_diversity', '244': 'eden_vectorizer', '245': 'sklearn_estimator_attributes', '246': 'Convert characters1', '247': 'minfi_getbeta', '248': 'cshl_word_list_grep', '249': 'melt', '250': 'trimmomatic', '251': 'picard_FastqToSam', '252': 'bed_to_bigwig', '253': 'msnbase-filter-merge-msms', '254': 'ip_threshold', '255': 'FROGS_clustering', '256': 'fastq_filter', '257': 'stoceps_filteringsp', '258': 'abims_xcms_xcmsSet', '259': 'ggplot2_heatmap2', '260': 'mothur_dist_seqs', '261': 'hicexplorer_hicplotmatrix', '262': 'vcftools_annotate', '263': 'mothur_heatmap_sim', '264': 'replace_chromosome_names', '265': 'cshl_fastx_nucleotides_distribution', '266': 'param_value_from_file', '267': 'OpenSwathWorkflow', '268': 'mothur_amova', '269': 'cshl_sed_tool', '270': 'locarna_best_subtree', '271': 'seqtk_trimfq', '272': 'raceid_inspecttrajectory', '273': 'varscan_copynumber', '274': 'picard_ValidateSamFile', '275': 'wc_gnu', '276': 'megahit_contig2fastg', '277': 'trinotate', '278': 'cardinal_mz_images', '279': 'compose_text_param', '280': 'hcluster_sg_parser', '281': 'ip_overlay_segmentation', '282': 'tp_sort_rows', '283': 'scanpy_scale_data', '284': 'picard_CollectInsertSizeMetrics', '285': 'ip_viz_overlay_moving_and_fixed_image', '286': 'cuffdiff', '287': 'gops_cluster_1', '288': 'viennarna_rnaplot', '289': 'treebest_best', '290': 'mtbls520_19b_seasons_unique', '291': 'uchime', '292': 'ctb_ob_svg_depiction', '293': 'FidoAdapter', '294': 'vegan_diversity', '295': 'meme_fimo', '296': 'cmFinder', '297': 'sort1', '298': 'ctb_np-likeness-calculator', '299': 'mothur_sub_sample', '300': 'salmon', '301': 'glimmer_build-icm', '302': 'seq_composition', '303': 'uparse', '304': 'tp_awk_tool', '305': 'bcftools_consensus', '306': 'bio3d_rmsf', '307': 'tmhmm2', '308': 'Extract_RNA_seq_Evidence', '309': 'get_read_pipeline', '310': 'flexbar_split_RR_bcs', '311': '__TAG_FROM_FILE__', '312': 'cufflinks', '313': 'snpEff_build_gb', '314': 'fastq_paired_end_splitter', '315': 'rcas', '316': 'trimmer', '317': 'collection_column_join', '318': 'disco', '319': 'gff2bed1', '320': 'vcfannotate', '321': 'CONVERTER_interval_to_bedstrict_0', '322': 'QCCalculator', '323': 'hca_matrix_downloader', '324': 'CONVERTER_bed_to_gff_0', '325': 'ip_coordinates_of_roi', '326': 'OpenSwathDecoyGenerator', '327': 'Grep1', '328': 'secure_hash_message_digest', '329': 'openms_protein_quantifier', '330': 'snpSift_geneSets', '331': 'mothur_taxonomy_to_krona', '332': 'join_files_on_column_fuzzy', '333': 'abims_CAMERA_annotateDiffreport', '334': 'get_pdb', '335': 'rbc_mafft', '336': 'je_demultiplex', '337': 'cuffnorm', '338': 'edger', '339': 'ruvseq', '340': 'predict_pipeline', '341': 'velveth', '342': 'ctb_remDuplicates', '343': 'deeptools_compute_matrix', '344': 'gatk2_base_recalibrator', '345': 'structure_to_gspan', '346': 'xcms-dilutionfilter', '347': 'ctb_silicos_qed', '348': 'samtool_filter2', '349': 'FilterBAM', '350': 'align_families', '351': 'goenrichment', '352': 'createInterval', '353': 'dexseq_annotate', '354': 'gatk2_depth_of_coverage', '355': 'samtools_filter', '356': 'deeptools_plot_fingerprint', '357': 'psy_maps', '358': 'deeptools_plot_pca', '359': 'mothur_align_seqs', '360': 'EMBOSS: fuzzpro38', '361': 'samtools_depth', '362': 'gatk2_reduce_reads', '363': 'bowtie2', '364': 'mothur_count_seqs', '365': 'vcffilter', '366': 'raceid_filtnormconf', '367': 'ggplot2_histogram', '368': 'gtftobed12', '369': 'gffcompare_to_bed', '370': 'ctb_obgrep', '371': 'ip_binary_to_labelimage', '372': 'jbrowse', '373': 'svm_classifier', '374': 'mimodd_map', '375': '__ZIP_COLLECTION__', '376': 'viz_overlay_moving_and_fixed_image', '377': 'tp_find_and_replace', '378': 'mothur_classify_rf', '379': 'vt_normalize', '380': 'AnnovarShed', '381': 'msgfplus', '382': 'myrimatch', '383': 'maldi_quant_peak_detection', '384': 'vcf2tsv', '385': 'numeric_clustering', '386': 'mtbls520_19e_seasons_features', '387': 'cshl_cut_tool', '388': 'metfrag-cli-batch-multiple', '389': 'sklearn_build_pipeline', '390': 'ip_2d_filter_segmentation_by_features', '391': 'htseq-count', '392': 'prokka', '393': 'metaquantome_expand', '394': 'hicexplorer_hictransform', '395': 'snpfinder', '396': 'rna_starsolo', '397': 'maldi_quant_preprocessing', '398': 'blast_reciprocal_best_hits', '399': 'FeatureLinkerUnlabeled', '400': 'mothur_get_groups', '401': 'tp_easyjoin_tool', '402': 'wolf_psort', '403': 'keras_model_builder', '404': 'EMBOSS: newseq59', '405': 'volcanoplot', '406': 'mothur_trim_seqs', '407': 'heinz_visualization', '408': 'ctb_pubchem_download_as_smiles', '409': 'kraken-filter', '410': 'fastp', '411': 'bg_column_arrange_by_header', '412': 'deeptools_bamCompare', '413': 'cshl_sort_header', '414': 'cuffmerge', '415': 'rseqc_insertion_profile', '416': 'CONVERTER_fasta_to_tabular', '417': 'scanpy_read_10x', '418': 'tp_unfold_column_tool', '419': 'docking', '420': 'IDMerger', '421': 'filter_columns', '422': 'mothur_get_communitytype', '423': 'cuffcompare', '424': 'Remove beginning1', '425': 'umi_tools_count', '426': 'FeatureFinderMultiplex', '427': 'qiime_split_libraries', '428': 'mass_spectrometry_imaging_segmentations', '429': 'CONVERTER_bam_to_bigwig_0', '430': 'fastq_manipulation', '431': 'deeptools_computeMatrix', '432': 'gtf_filter_by_attribute_values_list', '433': 'map_peptides_to_bed', '434': 'smf_utils_find-boxes', '435': 'XTandemAdapter', '436': 'cshl_grep_tool', '437': 'graphlan_annotate', '438': '__FILTER_FROM_FILE__', '439': 'miranda', '440': 'mtbls520_12_species_marchantia', '441': 'velvetg_jgi', '442': 'IDConflictResolver', '443': 'graphclust_aggregate_alignments', '444': 'Kneeplot', '445': 'qiime_align_seqs', '446': 'qual_stats_boxplot', '447': 'mothur_nmds', '448': 'mothur_heatmap_bin', '449': 'bedtools_complementbed', '450': 'rseqc_FPKM_count', '451': 'express', '452': 'EMBOSS: extractseq35', '453': 'CONVERTER_fastqgz_to_fastq', '454': 'gspan', '455': 'fragmenter', '456': 'humann2_unpack_pathways', '457': 'EMBOSS: revseq82', '458': 'kallisto_quant', '459': 'heinz', '460': 'blast2go', '461': 'gd_calc_freq', '462': 'cshl_fastx_trimmer', '463': '_ensembl_gtf2gene_list', '464': 'sailfish', '465': 'mummer_mummer', '466': 'stringtie_merge', '467': 'limma_voom', '468': 'openms_id_mapper', '469': 'CONVERTER_Bam_Bai_0', '470': 'snippy_core', '471': 'proteomics_moff', '472': 'gatk2_variant_select', '473': 'samtools_bedcov', '474': 'cshl_fastx_artifacts_filter', '475': 'flye', '476': 'mothur_phylotype', '477': 'bcftools_filter', '478': 'mergeCols1', '479': 'xcms-blankfilter', '480': 'decoyfasta', '481': 'mtbls520_19a_seasons_shannon', '482': 'Extract genomic DNA 1', '483': 'cast', '484': 'macs2_predictd', '485': 'infernal_cmstat', '486': 'csv_to_tabular', '487': 'varscan_mpileup', '488': 'seurat_run_pca', '489': 'scanpy_compute_graph', '490': 'metaphlan2', '491': 'regex1', '492': 'qiime_pick_open_reference_otus', '493': 'search_gui', '494': 'hicexplorer_hicmergematrixbins', '495': 'muscle', '496': 'camera-group-fwhm', '497': 'clusterprofiler_go', '498': 'sam_pileup', '499': 'mothur_unique_seqs', '500': 'gemini_query', '501': 'ctb_change_title', '502': 'methtools_plot', '503': 'bedtools_bed12tobed6', '504': 'Show tail1', '505': 'picard_ARRG', '506': 'vcfflatten2', '507': 'mothur_make_design', '508': 'Heatmap', '509': 'maker', '510': 'samtools_sort', '511': 'vegan_rarefaction', '512': 'rseqc_mismatch_profile', '513': 'FileFilter', '514': 'busco', '515': 'CONVERTER_bedgraph_to_bigwig', '516': 'sam2interval', '517': 'fastq_to_fasta_python', '518': 'bg_diamond', '519': 'openbabel_compound_convert', '520': 'PeakPickerHiRes', '521': 'lca1', '522': 'Univariate', '523': 'smf_utils_estimate-energy', '524': 'rbc_mirdeep2', '525': 'minfi_rset', '526': 'resistome_analyzer', '527': 'deeptools_alignmentsieve', '528': 'heinz_bum', '529': 'ip_projective_transformation_points', '530': 'gmx_em', '531': 'gatk2_print_reads', '532': 'rsem_calculate_expression', '533': 'circgraph', '534': 'picard_SamToFastq', '535': 'fgsea', '536': 'CONVERTER_sam_to_unsorted_bam', '537': 'flexbar_no_split', '538': 'seq_filter_by_mapping', '539': 'msconvert_nix', '540': 'ncbi_blastp_wrapper', '541': 'blastxml_to_top_descr', '542': 'venn_list', '543': 'EMBOSS: tranalign100', '544': 'prepare_ligand', '545': '__BUILD_LIST__', '546': 'deseq2', '547': 'sklearn_searchcv', '548': 'samtools_idxstats', '549': 'vcffilter2', '550': 'bismark_deduplicate', '551': 'bwameth', '552': 'blockclust', '553': '__APPLY_RULES__', '554': 'IDFilter', '555': 'minfi_getM', '556': 'stacks_denovomap', '557': 'tp_head_tool', '558': 'regionalgam_plot_trend', '559': 'react_cal_pipeline', '560': 'NFS_transfer', '561': 'seurat_scale_data', '562': 'mothur_summary_single', '563': 'CONVERTER_interval_to_bgzip_0', '564': 'mass_spectrometry_imaging_preprocessing', '565': 'vcftools_slice', '566': 'cshl_fastx_clipper', '567': 'Paste1', '568': 'hgv_linkToGProfile', '569': 'idr-embl', '570': 'picard_EstimateLibraryComplexity', '571': 'column_order_header_sort', '572': 'rnaz', '573': 'varscan_somatic', '574': 'heatmapper', '575': 'prepare_box', '576': 'mimodd_varreport', '577': 'valet', '578': 'fastq_join', '579': 'blockbuster', '580': 'gops_intersect_1', '581': 'collapse_dataset', '582': 'bamtools', '583': 'msconvert2_raw', '584': 'scale_image', '585': 'hisat2', '586': 'ctb_spectrophore_search', '587': 'aggregate_scores_in_intervals2', '588': 'rseqc_geneBody_coverage', '589': 'diffbind', '590': 'seurat_normalise_data', '591': 'stacks_refmap', '592': 'BAMTagHistogram', '593': 'align_back_trans', '594': 'bandage_image', '595': 'convert_bc_to_binary_RY.py', '596': 'translate_bed', '597': 'cshl_awk_replace_in_column', '598': 'IDPosteriorErrorProbability', '599': '__EXTRACT_DATASET__', '600': 'ctb_multi_obgrep', '601': 'CONVERTER_bam_to_coodinate_sorted_bam', '602': 'tophat', '603': 'datamash_reverse', '604': 'basil', '605': '__SORTLIST__', '606': 'htseq_count', '607': 'je_clip', '608': 'mtbls520_04_preparations', '609': 'genrich', '610': 'FileConverter', '611': 'raceid_inspectclusters', '612': 'samtools_slice_bam', '613': 'vcfvcfintersect', '614': 'snpSift_filter', '615': 'methtools_calling', '616': 'locarna_multiple', '617': 'goslimmer', '618': 'pipmir', '619': 'silac_analyzer', '620': 'mtbls520_23_seasons_rda', '621': 't2t_report', '622': 'extract_bcs.py', '623': 'enhanced_bowtie_wrapper', '624': 'minfi_read450k', '625': 'gemini_comp_hets', '626': 'term_id_vs_term_def', '627': 'seqtk_mergepe', '628': 'flanking_features_1', '629': 'plotly_parallel_coordinates_plot', '630': 'rmcontamination', '631': 'scanpy_filter_cells', '632': 'plotly_regression_performance_plots', '633': 'minfi_dropsnp', '634': 'thermo_raw_file_converter', '635': 'mothur_classify_seqs', '636': 'TrimByN', '637': 'FeatureFinderMetabo', '638': 'hgv_snpFreq', '639': 'ProteinInference', '640': 'cardinal_spectra_plots', '641': 'filter_tabular', '642': 'bedtools_bedtobam', '643': 'blastxml_to_tabular_selectable', '644': 'fastqc', '645': 'tp_grep_tool', '646': 'gemini_recessive_and_dominant', '647': 'qiime_make_otu_table', '648': 'bedtools_intersectBed', '649': 'seq_filter_by_id', '650': 'minfi_maptogenome', '651': 'mothur_homova', '652': 'ctb_online_data_fetch', '653': 'deg_annotate', '654': 'velvet', '655': 'find_in_reference', '656': 'genbank_to_gff', '657': 'bismark_bowtie', '658': 'flexbar', '659': 'get_child_terms', '660': 'ctb_chemfp_butina_clustering', '661': 'rseqc_bam_stat', '662': 'lastz_wrapper_2', '663': 'egsea', '664': 'regionalgam_autocor_acf', '665': 'translate_bed_sequences', '666': 'proteomics_search_peptide_prophet_1', '667': 'ncbi_rpsblast_wrapper', '668': 'ete_genetree_splitter', '669': 'bedtools_unionbedgraph', '670': 'samtools_split', '671': 'sklearn_feature_selection', '672': 'velvetg', '673': 'retrieve_scxa', '674': 'bedtools_windowbed', '675': 'FROGS_preprocess', '676': 'cshl_fasta_formatter', '677': 'vcfsort', '678': 'IDMassAccuracy', '679': 'gemini_interactions', '680': 'gemini_load', '681': 'peakcalling_macs14', '682': 'picard_SortSam', '683': 'seurat_dim_plot', '684': 'mothur_parsimony', '685': 'histogram_rpy', '686': 'openms_id_file_converter', '687': 'create_or_update', '688': 'build_ucsc_custom_track_1', '689': 'remove_tail.py', '690': 'crosscontamination_barcode_filter', '691': 'CONVERTER_gz_to_uncompressed', '692': 'snpSift_dbnsfp_generic', '693': 'genomespace_exporter', '694': 'piranha', '695': 'bcftools_mpileup', '696': 'humann2', '697': 'cuffquant', '698': 'gatk_depth_of_coverage', '699': 'trinity_define_clusters_by_cutting_tree', '700': 'kraken-report', '701': 'macs2_callpeak', '702': 'Interval_Maf_Merged_Fasta2', '703': 'ip_2d_split_binaryimage_by_watershed', '704': 'snpsift_vartype', '705': 'pygenomeTracks', '706': 'imagecoordinates_flipaxis', '707': 'fastq_stats', '708': 'seurat_find_markers', '709': 'FileInfo', '710': 'minimap2', '711': 'vcfselectsamples', '712': 'custom_pro_db', '713': 'filtlong', '714': 'DatamashOps', '715': 'gmx_nvt', '716': 'tp_multijoin_tool', '717': 'prokaryotic_ncbi_submission', '718': 'flaimapper', '719': 'proteomics_search_protein_prophet_1', '720': 'ggplot2_heatmap', '721': 'nspdk_sparse', '722': 'fastq_combiner', '723': 'bg_find_subsequences', '724': 'sample_seqs', '725': 'gops_concat_1', '726': 'mothur_chimera_uchime', '727': 'raceid_clustering', '728': 'bedtools_coveragebed', '729': 'picard_CleanSam', '730': 'prinseq', '731': 'tsne', '732': 'bedtools_genomecoveragebed_histogram', '733': 'seqtk_mergefa', '734': 'wiggle2simple1', '735': 'mothur_split_abund', '736': 'sklear_numeric_clustering', '737': 'mlst', '738': 'flashlfq', '739': 'w4mjoinpn', '740': 'camera-find-isotopes', '741': 'addName', '742': 'picard_MergeSamFiles', '743': 'bamleftalign', '744': 'krona-text', '745': 'gops_join_1', '746': 'rbc_splitfasta', '747': 'checkFormat', '748': 'gatk2_unified_genotyper', '749': 'ucsc-twobittofa', '750': 'mass_spectrometry_imaging_classification', '751': 'spocc_occ', '752': 'gemini_de_novo', '753': 'rseqc_read_distribution', '754': 'rseqc_clipping_profile', '755': 'bg_statistical_hypothesis_testing', '756': 'ctb_confab', '757': 'nonpareil', '758': 'cshl_easyjoin', '759': 'deeptools_plot_enrichment', '760': 'Show beginning1', '761': 'prepare_ligands_for_docking', '762': 'picard_CollectRnaSeqMetrics', '763': 'p_clip_peaks', '764': 'mothur_filter_seqs', '765': 'cardinal_combine', '766': 'poretools_extract', '767': 'CometAdapter', '768': 'rbc_mirdeep2_mapper', '769': 'iqtree', '770': 'column_remove_by_header', '771': 'mass_spectrometry_imaging_qc', '772': 'mtbls520_08d_concentration', '773': 'bamFilter', '774': 'mothur_cluster_split', '775': 'idpqonvert', '776': 'scpipe', '777': 'ctb_ob_addh', '778': 'scaffold2fasta', '779': 'sklearn_model_validation', '780': 'MapStatistics', '781': 'viennarna_rnafold', '782': 'gemini_actionable_mutations', '783': 'pieplots_macs', '784': 'hicexplorer_hicbuildmatrix', '785': 'bctools_remove_tail', '786': 'bigwig_to_bedgraph', '787': 'Psortb', '788': 'peakachu', '789': 'gd_rank_terms', '790': 'porechop', '791': 'iReport', '792': 'rseqc_infer_experiment', '793': 'graphicsmagick_image_montage', '794': 'gatk_realigner_target_creator', '795': 'seurat_run_tsne', '796': 'pathview', '797': 'FeatureLinkerUnlabeledQT', '798': 'methtools_filter', '799': 'Transformation', '800': 'bwa', '801': 'mass_spectrometry_imaging_filtering', '802': 'rseqc_RNA_fragment_size', '803': 'mimodd_header', '804': 'feelnc', '805': 'deeptools_multi_bam_summary', '806': 'tp_replace_in_line', '807': 'Batch_correction', '808': 'msconvert3_raw', '809': 'camera-annotate-peaks', '810': 'bamCoverage_deepTools', '811': 'minfi_ppfun', '812': 'ctb_rdkit_describtors', '813': 'cor2', '814': 'tp_sorted_uniq', '815': 'filter_bed_on_splice_junctions', '816': 'velveth_jgi', '817': 'gatk_indel_realigner', '818': 'vcftools_compare', '819': 'bcftools_stats', '820': 'mtbls520_05a_import_maf', '821': 'vcftools_merge', '822': 'trinity_align_and_estimate_abundance', '823': 'mtbls520_19d_seasons_concentration', '824': 'bamCompare_deepTools', '825': 'ctb_opsin', '826': 'mummer_nucmer', '827': 'bedtools_getfastabed', '828': 'CONVERTER_bed_gff_or_vcf_to_bigwig_0', '829': 'mtbls520_08a_species_shannon', '830': 'regionalgam_gls', '831': 'rseqc_junction_annotation', '832': 'chipeakanno_annopeaks', '833': 'sickle', '834': 'ncbi_makeblastdb', '835': 'mothur_classify_otu', '836': 'ProteinResolver', '837': 'picard_CollectWgsMetrics', '838': 'IDMapper', '839': 'umi_tools_group', '840': 'vigiechiro_idvalid', '841': 'gffread', '842': 'flexbardsc', '843': 'sklearn_numeric_clustering', '844': 'picard_CollectGcBiasMetrics', '845': 'bams2ratio', '846': 'ggplot_histogram', '847': 'mycrobiota-correct-replicates', '848': 'ctb_remove_protonation_state', '849': 'export2graphlan', '850': 'metaspades', '851': 'deeptools_compute_gc_bias', '852': 'scanpy_find_cluster', '853': 'ip_filter_standard', '854': 'extract_aln_ends.py', '855': 'seurat_create_seurat_object', '856': 'cardinal_filtering', '857': 'unicycler', '858': 'graphclust_align_cluster', '859': 'sklearn_nn_classifier', '860': 'bedtools_bamtofastq', '861': 'minfi_qc', '862': 'samtools_stats', '863': 'get_sequences', '864': 'annotateMyIDs', '865': 'ip_imageinfo', '866': 'Draw_phylogram', '867': 'vsearch_dereplication', '868': 'ctb_filter', '869': 'je_markdupes', '870': 'mothur_unifrac_unweighted', '871': 'metilene', '872': 'Fetch Taxonomic Ranks', '873': 'MSGFPlusAdapter', '874': 'deeptools_correct_gc_bias', '875': 'minfi_getanno', '876': 'rbc_rnacode', '877': 'rawtools', '878': 'MapRTTransformer', '879': 'paralyzer', '880': 'xcms_merge', '881': 'bismark_bowtie2', '882': 'cshl_uniq_tool', '883': 'seurat_find_clusters', '884': 'rarefaction_analyzer', '885': 'gtf2bedgraph', '886': 'tophat2', '887': 'gatk2_variant_annotator', '888': 'proteomics_search_tandem_1', '889': 'massbank_ws_searchspectrum', '890': 'deseq2_single', '891': 'peakcalling_macs', '892': 'IDFileConverter', '893': 'mothur_make_biom', '894': 'ip_count_objects', '895': 'cshl_fastx_reverse_complement', '896': 'mothur_summary_seqs', '897': 'infernal_cmbuild', '898': 'EMBOSS: water107', '899': 'w4mclassfilter', '900': 'bedtools_coveragebed_counts', '901': 'bedtools_genomecoveragebed', '902': 'vcfallelicprimitives', '903': 'picard_CASM', '904': 'rseqc_read_duplication', '905': 'addValue', '906': 'glimmer_knowlegde-based', '907': 'intarna', '908': 'ctb_chemfp_mol2fps', '909': 'rnafold', '910': 'qiime_pick_closed_reference_otus', '911': 'graphlan', '912': 'gd_raxml', '913': 'msnbase-read-msms', '914': 'heatmap2', '915': 'EMBOSS: textsearch98', '916': 'snpeff_to_peptides', '917': 'passatutto', '918': 'picard_QualityScoreDistribution', '919': 'xcms-collect-peaks', '920': 'roary', '921': 'meme_dreme', '922': 'msms2metfrag-multiple', '923': 'gemini_burden', '924': 'qiime_make_phylogeny', '925': 'plotly_ml_performance_plots', '926': 'fasta_compute_length', '927': 'md_converter', '928': 'gbk_to_orf', '929': 'tp_tail_tool', '930': 'trinity_analyze_diff_expr', '931': 'glob_report', '932': 'extract_element_from_collection', '933': 'trinity_abundance_estimates_to_matrix', '934': 'mycrobiota-krona-mothur', '935': 'length_and_gc_content', '936': 'vcfgenotypes', '937': 'fasttree', '938': 'hisat', '939': 'cds_essential_variability', '940': 'pileup_interval', '941': '__RELABEL_FROM_FILE__', '942': 'qiime_pick_otus', '943': 'freebayes_wrapper', '944': 'quality_metrics', '945': 'scanpy_find_markers', '946': 'hicexplorer_hicsummatrices', '947': 'IDScoreSwitcher', '948': 'picard_ReorderSam', '949': 'bctools_extract_alignment_ends', '950': 'gemini_db_info', '951': 'gff_filter_by_attribute', '952': 'abims_xcms_fillPeaks', '953': 'EMBOSS: geecee41', '954': 'gmx_md', '955': 'clusterprofiler_bitr', '956': 're_he_maldi_image_registration', '957': 'rnbeads', '958': 'deeptools_plot_correlation', '959': 'mothur_corr_axes', '960': 'gatk2_variant_apply_recalibration', '961': 'humann2_renorm_table', '962': 'mothur_make_group', '963': 'umi_tools_extract', '964': 'minfi_dmr', '965': 'heinz_scoring', '966': 'ip_landmark_registration', '967': 'interproscan', '968': 'ExtractFASTAfromFASTQ', '969': 'TrimPrimer', '970': 'ssake', '971': 'Extract_features1', '972': 'featurecounts', '973': 'wig_to_bigWig', '974': 'FalseDiscoveryRate', '975': 'bedtools_map', '976': 'gemini_inheritance', '977': 'PlasFlow', '978': 'bam-filter', '979': 'ip_scale_image', '980': 'smooth_running_window', '981': 'nn_classifier', '982': 'cshl_find_and_replace', '983': 'Flash', '984': 'nanoplot', '985': 'varscan', '986': 'cummerbund_to_cuffdiff', '987': 'ProteinQuantifier', '988': 'cmv', '989': 'mothur_chimera_vsearch', '990': 'bundle_collection', '991': 'smf_utils_fix-fasta-headers', '992': 'fastq_trimmer', '993': 'gdal_ogr2ogr', '994': 'ete_homology_classifier', '995': 'transpose', '996': 'ctb_rdkit_descriptors', '997': 'vcffixup', '998': 'uniprot', '999': 'nanopolish_variants', '1000': 'bedtools_bamtobed', '1001': 'ia_coordinates_of_roi', '1002': 'sam_to_bam', '1003': 'cshl_fastq_quality_filter', '1004': 'mothur_rarefaction_single', '1005': 'deeptools_bigwigCompare', '1006': 'Summary_Statistics1', '1007': 'flexbar_split_RYYR_bcs', '1008': 'multiqc', '1009': 'eukaryotic_ncbi_submission', '1010': 'hicexplorer_hiccorrectmatrix', '1011': 'mothur_merge_files', '1012': 'replace_column_with_key_value_file', '1013': 'bam_to_bigwig', '1014': 'megablast_wrapper', '1015': 'bio3d_pca', '1016': 'cardinal_quality_report', '1017': 'megablast_xml_parser', '1018': 'bed_to_bigBed', '1019': 'vcf_filter', '1020': 'merge_pcr_duplicates.py', '1021': 'mothur_screen_seqs', '1022': 'picard_NormalizeFasta', '1023': 'hicexplorer_hicplottads', '1024': 'PicardInsertSize', '1025': 'heatmapper_deepTools', '1026': 'deeptools_profiler', '1027': 'gatk_unified_genotyper', '1028': 'gmx_setup', '1029': 'cd_hit_est', '1030': 'gene2exon1', '1031': 'igvtools_count', '1032': 'hcluster_sg', '1033': 'seqtk_sample', '1034': 'qiime_pick_rep_set', '1035': 'bctools_extract_crosslinked_nucleotides', '1036': 'annotatemyids', '1037': 'rgPicFixMate', '1038': 'bed2gff1', '1039': 'gops_merge_1', '1040': 'sklearn_ensemble', '1041': 'bedtools_groupbybed', '1042': 'CONVERTER_bed_to_fli_0', '1043': 'dt_profiler', '1044': 'combine_metaphlan2_humann2', '1045': 'macs2_bdgcmp', '1046': 'antismash', '1047': 'bwa_wrapper', '1048': 'FastTree', '1049': 'biosigner', '1050': 'seurat_find_variable_genes', '1051': 'mycrobiota-otutable_add_blast', '1052': 'FeatureFinderCentroided', '1053': 'Count1', '1054': 'idpassemble', '1055': 'bedtools_shufflebed', '1056': 'xcms-group-peaks', '1057': 'cardinal_classification', '1058': 'regionalgam_glmmpql', '1059': 'FROGS_filters', '1060': 'augustus_training', '1061': 'rseqc_inner_distance', '1062': 'tp_replace_in_column', '1063': 'iuc_pear', '1064': 'allele_counts_1', '1065': 'scanpy_parameter_iterator', '1066': 'sklearn_clf_metrics', '1067': 'rm_spurious_events.py', '1068': 'CONVERTER_gff_to_bed_0', '1069': 'velvetoptimiser', '1070': 'metaquantome_stat', '1071': 'snpEff', '1072': 'bowtie_wrapper', '1073': 'sam_merge2', '1074': 'random_lines1', '1075': 'msconvert', '1076': 'cardinal_data_exporter', '1077': 'mummer_dnadiff', '1078': 'resize_coordinate_window', '1079': 'mycrobiota-split-multi-otutable', '1080': 'kraken-translate', '1081': 'metaphlan2krona', '1082': 'ctb_subsearch', '1083': 'bamSplit', '1084': 'chipsequtil_maptoknowngenes', '1085': 'generic_filter', '1086': 'Filter1', '1087': 'vigiechiro_bilanenrichipf', '1088': 'mothur_remove_lineage', '1089': 'mtbls520_10_species_varpart', '1090': 'shovill', '1091': 'methtools_dmr', '1092': 'correctGCBias', '1093': 'humann2_regroup_table', '1094': 'deeptools_bamFingerprint', '1095': 'mothur_pcoa', '1096': 'MAF_To_Fasta1', '1097': 'mothur_metastats', '1098': 'metams_runGC', '1099': 'camera-find-adducts', '1100': 'sm_mirdeep2core_without_Randfold', '1101': 'seurat_read10x', '1102': 'snpSift_extractFields', '1103': 'fastq_paired_end_deinterlacer', '1104': 'spades', '1105': 'mothur_venn', '1106': 'mothur_lefse', '1107': 'minfi_getsnp', '1108': 'deeptools_bam_pe_fragmentsize', '1109': 'samtools_flag_filter', '1110': 'mtbls520_09_species_venn', '1111': 'segemehl', '1112': 'fastq_groomer', '1113': 'aragorn_trna', '1114': 'raceid_main', '1115': 'Nucleosome', '1116': 'blast_parser', '1117': 'Grouping1', '1118': 'deeptools_multi_bigwig_summary', '1119': 'trinity', '1120': 'mass_spectrometry_imaging_ion_images', '1121': 'snpSift_int', '1122': 'mimodd_varcall', '1123': 'scanpy_normalise_data', '1124': 'FileMerger', '1125': 'ip_histogram_equalization', '1126': 'bismark_pretty_report', '1127': 'goseq', '1128': 'cshl_fastx_collapser', '1129': 'trinity_samples_qccheck', '1130': 'table_annovar', '1131': 'mothur_make_contigs', '1132': 'preMloc', '1133': 'ctb_remIons', '1134': 'staramr_search', '1135': 'deeptools_computeGCBias', '1136': 'RNAlien', '1137': 'ncbi_blastx_wrapper', '1138': 'Cut1', '1139': 'wtdbg', '1140': 'TextExporter', '1141': 'bcftools_plugin_counts', '1142': 'qiime_split_libraries_fastq', '1143': 'vigiechiro_bilanenrichirp', '1144': 'picard_FixMateInformation', '1145': 'sequence_content_trimmer', '1146': 'CONVERTER_fasta_to_2bit', '1147': 'get_feature_info', '1148': 'kraken2', '1149': 'tag_stat2', '1150': 'bedtools_annotatebed', '1151': 'pileup_parser', '1152': 'ctb_osra', '1153': 'tp_cat', '1154': 'mtbls520_16_ecology_rda', '1155': 'cummeRbund', '1156': 'maxquant', '1157': 'umi_tools_dedup', '1158': 'gatk2_indel_realigner', '1159': 'gmx_npt', '1160': 'pileometh', '1161': 'mimodd_vcf_filter', '1162': 'sam_bw_filter', '1163': 'FROGS_remove_chimera', '1164': 'eden_sequence_converter', '1165': 'deeptools_bam_coverage', '1166': 'tp_cut_tool', '1167': 'kraken', '1168': 'gemini_stats', '1169': 'freebayes', '1170': 'unzip', '1171': 'stoceps_glm', '1172': 'naive_variant_caller', '1173': 'metaquantome_viz', '1174': 'aggregate', '1175': 'vcfcombine', '1176': 'mothur_split_groups', '1177': 'snpSift_annotate', '1178': 'stacks_procrad', '1179': 'meme_psp_gen', '1180': 'hicexplorer_hicplotviewpoint', '1181': 'gemini_pathways', '1182': 'picard_MarkDuplicates', '1183': 'pilon', '1184': 'MetaProSIP', '1185': 'ngsutils_bam_filter', '1186': 'vigiechiro_idcorrect_2ndlayer', '1187': 'tp_sed_tool', '1188': 'cardinal_segmentations', '1189': 'ip_projective_transformation', '1190': 'mtbls520_06_import_traits', '1191': 'DatamashTranspose', '1192': 'cshl_fastx_renamer', '1193': 'cutadapt', '1194': 'seurat', '1195': 'Datamash', '1196': 'gmx_merge_topology_files', '1197': 'minfi_mset', '1198': 'mothur_unifrac_weighted', '1199': 'subtract_query1', '1200': 'rseqc_bam2wig', '1201': 'peptide_shaker', '1202': 'mothur_tree_shared', '1203': 'get_flanks1', '1204': 'quast', '1205': 'samtools_calmd', '1206': 'bedtools_subtractbed', '1207': 'ncbi_blastn_wrapper', '1208': 'scanpy_find_variable_genes', '1209': 'query_tabular', '1210': 'scanpy_filter_genes', '1211': 'picard_BamIndexStats', '1212': 'fastq collapser', '1213': 'TagBamWithReadSequenceExtended', '1214': 'metaphlan_hclust_heatmap', '1215': 'vcf2pgSnp', '1216': 'xpath', '1217': 'ctb_stripit', '1218': 'tabular_to_csv', '1219': 'tp_text_file_with_recurring_lines', '1220': 'bismark_methylation_extractor', '1221': '__FILTER_FAILED_DATASETS__', '1222': 'mlocarna', '1223': 'scanpy_run_umap', '1224': 'FeatureLinkerUnlabeledKD', '1225': 'cshl_fastq_quality_boxplot', '1226': 'regionalgam_ab_index', '1227': 'fraggenescan', '1228': 'format_metaphlan2_output', '1229': 'EMBOSS: transeq101', '1230': 'dbbuilder', '1231': 'HighResPrecursorMassCorrector', '1232': 'ctb_alignit', '1233': 'bio3d_rmsd', '1234': 'samtools_flagstat', '1235': 'seurat_filter_cells', '1236': 'cshl_fastx_quality_statistics', '1237': 'tp_split_on_column', '1238': 'ConsensusMapNormalizer', '1239': 'raceid_trajectory', '1240': 'modencode_peakcalling_macs2', '1241': 'rgPicardMarkDups', '1242': '__UNZIP_COLLECTION__', '1243': 'collection_element_identifiers', '1244': 'bandage_info', '1245': 'deeptools_heatmapper', '1246': 'ococo', '1247': 'bctools_remove_spurious_events', '1248': 'iframe', '1249': 'dropletutils', '1250': 'qiime_filter_alignment', '1251': 'transdecoder', '1252': 'calc_fitness', '1253': 'regex_replace', '1254': 'ChangeCase', '1255': 'bedtools_intersectbed', '1256': 'ucsc_cell_browser', '1257': 'ConsensusID', '1258': 'samtools_rmdup', '1259': 'picard_AddOrReplaceReadGroups', '1260': 'Btrim64', '1261': 'group_humann2_uniref_abundances_to_go', '1262': 'ip_2d_feature_extraction', '1263': 'NSPDK_candidateClust', '1264': 'EMBOSS: wordcount109', '1265': 'tp_tac', '1266': 'filter_by_fasta_ids', '1267': 'fastq_to_tabular', '1268': 'gemini_annotate', '1269': 'mycrobiota-make-multi-otutable', '1270': 'ggplot_point', '1271': 'eggnog_mapper', '1272': 'snpEff_download', '1273': 'sed_stream_editor', '1274': 'modencode_peakcalling_spp', '1275': 'EMBOSS: seqret84', '1276': 'CONVERTER_interval_to_bed_0', '1277': 'methtools_destrand', '1278': 'hgv_david', '1279': 'correct_barcodes', '1280': 'coords2clnt.py', '1281': 'canu'}\n"
     ]
    }
   ],
   "source": [
    "#base_path = \"data/rnn_custom_loss/complete_training/\"\n",
    "base_path = \"data/models/\"\n",
    "\n",
    "path_data_dict = base_path + \"data_dict.txt\"\n",
    "path_usage_wt = base_path + \"usage_prediction.txt\"\n",
    "path_class_wt = base_path + \"class_weights.txt\"\n",
    "path_test_data = base_path + \"test_paths_dict.txt\"\n",
    "model_path = base_path + \"trained_model_19_09.hdf5\"\n",
    "#class_weights = read_file(path_class_wt)\n",
    "\n",
    "\n",
    "#usage_weights = read_file(path_usage_wt)\n",
    "#data_dict = read_file(path_data_dict)\n",
    "trained_model = h5py.File(model_path, 'r')\n",
    "model_config = json.loads(trained_model.get('model_config').value)\n",
    "class_weights = json.loads(trained_model.get('class_weights').value)\n",
    "    \n",
    "loaded_model = model_from_json(model_config)\n",
    "dictionary = json.loads(trained_model.get('data_dictionary').value)\n",
    "compatibile_tools = json.loads(trained_model.get('compatible_tools').value)\n",
    "best_params = json.loads(trained_model.get('best_parameters').value)\n",
    "\n",
    "model, dictionary, reverse_dictionary, compatibile_tools = create_model(model_path)\n",
    "\n",
    "print(reverse_dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicted tools: \n",
      "\n",
      "{'shovill': 100, 'bwa_wrapper': 100, 'velvetoptimiser': 100, 'fastq_stats': 100, 'bismark_bowtie2': 99, 'fastp': 99, 'unicycler': 99, 'hisat2': 99, 'bowtie_wrapper': 99, 'tophat2': 99, 'trinity_align_and_estimate_abundance': 99, 'cutadapt': 99, 'prinseq': 99, 'mothur_make_contigs': 99, 'sickle': 99, 'fastq_trimmer': 99}\n",
      "\n",
      "Class weights: \n",
      "\n",
      "{}\n",
      "\n",
      "Usage weights: \n",
      "\n",
      "{}\n",
      "\n",
      "Mean usage wt: nan\n",
      "\n",
      "Inverted weights: \n",
      "\n",
      "{}\n",
      "\n",
      "Tool ids\n",
      "{'shovill': 1090, 'bwa_wrapper': 1047, 'velvetoptimiser': 1069, 'fastq_stats': 707, 'bismark_bowtie2': 881, 'fastp': 410, 'unicycler': 857, 'hisat2': 585, 'bowtie_wrapper': 1072, 'tophat2': 886, 'trinity_align_and_estimate_abundance': 822, 'cutadapt': 1193, 'prinseq': 730, 'mothur_make_contigs': 1131, 'sickle': 833, 'fastq_trimmer': 992}\n",
      "======================================\n"
     ]
    }
   ],
   "source": [
    "topk = 20\n",
    "tool_seq = \"250\"\n",
    "class_wt, usage_wt, inverse_wt, pred_tools = verify_model(model, tool_seq, \"\", dictionary, reverse_dictionary, compatibile_tools, class_weights, topk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
