{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import json\n",
    "import warnings\n",
    "import operator\n",
    "\n",
    "import h5py\n",
    "from keras.models import model_from_json\n",
    "from keras import backend as K\n",
    "from keras.utils import get_custom_objects\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "size_title = 18\n",
    "size_label = 14\n",
    "n_pred = 2\n",
    "\n",
    "base_path = \"data/evaluate_rnn_compatible_loss_current/\"\n",
    "\n",
    "path_data_dict = base_path + \"data_dict.txt\"\n",
    "path_usage_wt = base_path + \"usage_prediction.txt\"\n",
    "path_class_wt = base_path + \"class_weights.txt\"\n",
    "path_test_data = base_path + \"test_paths_dict.txt\"\n",
    "model_path = base_path + \"trained_model.hdf5\"\n",
    "\n",
    "def read_file(file_path):\n",
    "    with open(file_path, 'r') as data_file:\n",
    "        data = json.loads(data_file.read())\n",
    "    return data\n",
    "\n",
    "class_weights = read_file(path_class_wt)\n",
    "usage_weights = read_file(path_usage_wt)\n",
    "data_dict = read_file(path_data_dict)\n",
    "\n",
    "\n",
    "def output_act(class_weights):\n",
    "    weights = list(class_weights.values())\n",
    "    def weighted_sigmoid_activation(x):\n",
    "        exp_weights = K.expand_dims(weights, axis=-1)\n",
    "        return K.sigmoid(x * K.transpose(exp_weights))\n",
    "    return weighted_sigmoid_activation\n",
    "\n",
    "\n",
    "get_custom_objects().update({'weighted_sigmoid_activation': output_act(class_weights)})\n",
    "\n",
    "def create_model(model_path):\n",
    "    \n",
    "    trained_model = h5py.File(model_path, 'r')\n",
    "    model_config = json.loads(trained_model.get('model_config').value)\n",
    "    loaded_model = model_from_json(model_config, custom_objects={'activation': output_act(class_weights)})\n",
    "    dictionary = json.loads(trained_model.get('data_dictionary').value)\n",
    "    compatibile_tools = json.loads(trained_model.get('compatible_tools').value)\n",
    "    best_params = json.loads(trained_model.get('best_parameters').value)\n",
    "    reverse_dictionary = dict((str(v), k) for k, v in dictionary.items())\n",
    "    model_weights = list()\n",
    "    weight_ctr = 0\n",
    "    while True:\n",
    "        try:\n",
    "            d_key = \"weight_\" + str(weight_ctr)\n",
    "            weights = trained_model.get(d_key).value\n",
    "            model_weights.append(weights)\n",
    "            weight_ctr += 1\n",
    "        except Exception as exception:\n",
    "            break\n",
    "    # set the model weights\n",
    "    loaded_model.set_weights(model_weights)\n",
    "    return loaded_model, dictionary, reverse_dictionary, compatibile_tools\n",
    "\n",
    "model, dictionary, reverse_dictionary, compatibile_tools = create_model(model_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def analyze_output_layer(model, test_sample, dimensions, phase, iter_num=10):\n",
    "    output_last = K.function([model.layers[0].input, K.learning_phase()], [model.layers[-1].output])\n",
    "    result = np.zeros((iter_num,) + (1, dimensions))\n",
    "    for idx in range(iter_num):\n",
    "        result[idx] = output_last([test_sample, phase])[0]\n",
    "    prediction = result.mean(axis=0)\n",
    "    uncertainty = result.var(axis=0)\n",
    "    return prediction, uncertainty\n",
    "\n",
    "\n",
    "def plot_error_bar(prediction, variance):\n",
    "    pred = prediction[0].tolist()\n",
    "    var = variance[0].tolist()\n",
    "    x = range(0, len(pred))\n",
    "    plt.figure(figsize = (12, 12))\n",
    "    #plt.errorbar(x, pred, var, marker='s', mfc='red', mec='green', ms=20, mew=4)\n",
    "    plt.errorbar(x, pred, var, fmt='o', marker='s', mfc='red', mec='green')\n",
    "    plt.grid(True)\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def verify_model(model, tool_sequence, labels, dictionary, reverse_dictionary, compatible_tools, phase, topk=20, max_seq_len=25):\n",
    "    tl_seq = tool_sequence.split(\",\")\n",
    "    last_tool_name = reverse_dictionary[str(tl_seq[-1])]\n",
    "    last_compatible_tools = compatible_tools[last_tool_name]\n",
    "    sample = np.zeros(max_seq_len)\n",
    "    for idx, tool_id in enumerate(tl_seq):\n",
    "        sample[idx] = int(tool_id)\n",
    "    sample_reshaped = np.reshape(sample, (1, max_seq_len))\n",
    "\n",
    "    tool_sequence_names = [reverse_dictionary[str(tool_pos)] for tool_pos in tool_sequence.split(\",\")]\n",
    "    print(\"Tool seq: %s\" % \",\".join(tool_sequence_names))\n",
    "    \n",
    "    # predict next tools for a test path\n",
    "    prediction = model.predict(sample_reshaped, verbose=0)\n",
    "    #prediction, uncertainty = analyze_output_layer(model, sample_reshaped, len(reverse_dictionary) + 1, phase)\n",
    "    #plot_error_bar(prediction, uncertainty)\n",
    "    \n",
    "    prediction = np.reshape(prediction, (prediction.shape[1],))\n",
    "    prediction_pos = np.argsort(prediction, axis=-1)\n",
    "\n",
    "    # get topk prediction\n",
    "    topk_prediction_pos = prediction_pos[-topk:]\n",
    "    topk_prediction_val = [np.round(prediction[pos], 2) for pos in topk_prediction_pos]\n",
    "\n",
    "    # read tool names using reverse dictionary\n",
    "    pred_tool_ids = [reverse_dictionary[str(tool_pos)] for tool_pos in topk_prediction_pos]\n",
    "    actual_next_tool_ids = list(set(pred_tool_ids).intersection(set(last_compatible_tools.split(\",\"))))      \n",
    "\n",
    "    print()\n",
    "    pred_tool_ids_sorted = dict()\n",
    "    for (tool_pos, tool_pred_val) in zip(topk_prediction_pos, topk_prediction_val):\n",
    "        tool_name = reverse_dictionary[str(tool_pos)]\n",
    "        if tool_name in actual_next_tool_ids:\n",
    "            pred_tool_ids_sorted[tool_name] = tool_pred_val\n",
    "    pred_tool_ids_sorted = dict(sorted(pred_tool_ids_sorted.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    \n",
    "    cls_wt = dict()\n",
    "    usg_wt = dict()\n",
    "    inv_wt = dict()\n",
    "    ids_tools = dict()\n",
    "    keys = list(pred_tool_ids_sorted.keys())\n",
    "    for k in keys:\n",
    "        try:\n",
    "            cls_wt[k] = np.round(class_weights[str(data_dict[k])], 2)\n",
    "            usg_wt[k] = np.round(usage_weights[k], 2)\n",
    "            inv_wt[k] = np.round(inverted_weights[str(data_dict[k])], 2)\n",
    "        except:\n",
    "            continue\n",
    "    print(\"Predicted tools: \\n\")\n",
    "    print(pred_tool_ids_sorted)\n",
    "    print()\n",
    "    print(\"Class weights: \\n\")\n",
    "    cls_wt = dict(sorted(cls_wt.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    print(cls_wt)\n",
    "    print()\n",
    "    print(\"Usage weights: \\n\")\n",
    "    usg_wt = dict(sorted(usg_wt.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    print(usg_wt)\n",
    "    print()\n",
    "    total_usage_wt = np.mean(list(usg_wt.values()))\n",
    "    print(\"Mean usage wt: %0.4f\" % (total_usage_wt))\n",
    "    print()\n",
    "    print(\"Inverted weights: \\n\")\n",
    "    inv_wt = dict(sorted(inv_wt.items(), key=lambda kv: kv[1], reverse=True))\n",
    "    print(inv_wt)\n",
    "    for key in pred_tool_ids_sorted:\n",
    "        ids_tools[key] = dictionary[key]\n",
    "    print()\n",
    "    print(\"Tool ids\")\n",
    "    print(ids_tools)\n",
    "    print(\"======================================\")\n",
    "    return cls_wt, usg_wt, inv_wt, pred_tool_ids_sorted\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1': 'chipsequtil_maptoknowngenes',\n",
       " '2': 'mothur_summary_seqs',\n",
       " '3': 'Digestor',\n",
       " '4': 'Nucleosome',\n",
       " '5': 'gatk2_variant_apply_recalibration',\n",
       " '6': 'deeptools_plot_correlation',\n",
       " '7': 'snpeff_to_peptides',\n",
       " '8': '__ZIP_COLLECTION__',\n",
       " '9': 'rbc_mirdeep2_quantifier',\n",
       " '10': 'ggplot2_point',\n",
       " '11': 'qiime_make_phylogeny',\n",
       " '12': 'abims_CAMERA_annotateDiffreport',\n",
       " '13': 'metilene',\n",
       " '14': 'snippy_core',\n",
       " '15': 'mothur_tree_shared',\n",
       " '16': 'egsea',\n",
       " '17': 'mothur_lefse',\n",
       " '18': 'gatk2_base_recalibrator',\n",
       " '19': 'picard_MergeSamFiles',\n",
       " '20': 'blast2go',\n",
       " '21': 'tp_head_tool',\n",
       " '22': 'rbc_splitfasta',\n",
       " '23': 'bedtools_bamtofastq',\n",
       " '24': 'rnaz',\n",
       " '25': 'Add_a_column1',\n",
       " '26': 'methtools_dmr',\n",
       " '27': 'bcftools_view',\n",
       " '28': 'cuffnorm',\n",
       " '29': 'wig_to_bigWig',\n",
       " '30': 'Psortb',\n",
       " '31': 'mothur_classify_seqs',\n",
       " '32': 'cshl_grep_tool',\n",
       " '33': 'NSPDK_candidateClust',\n",
       " '34': 'krona-text',\n",
       " '35': 'Cut1',\n",
       " '36': 'FeatureLinkerUnlabeled',\n",
       " '37': 'cshl_fastx_artifacts_filter',\n",
       " '38': 'bedtools_shufflebed',\n",
       " '39': 'kraken-translate',\n",
       " '40': 'mothur_count_seqs',\n",
       " '41': 'IDScoreSwitcher',\n",
       " '42': 'picard_CollectInsertSizeMetrics',\n",
       " '43': 'uniprot',\n",
       " '44': 'ConsensusID',\n",
       " '45': 'Extract_features1',\n",
       " '46': 'gtf_filter_by_attribute_values_list',\n",
       " '47': 'wolf_psort',\n",
       " '48': 'mass_spectrometry_imaging_preprocessing',\n",
       " '49': 'qiime_pick_closed_reference_otus',\n",
       " '50': 'gff_to_sequence',\n",
       " '51': 'interproscan',\n",
       " '52': 'kraken-filter',\n",
       " '53': 'FROGS_affiliations_stat',\n",
       " '54': 'ip_projective_transformation',\n",
       " '55': 'mothur_amova',\n",
       " '56': 'datamash_transpose',\n",
       " '57': 'cardinal_spectra_plots',\n",
       " '58': 'sam_bw_filter',\n",
       " '59': 'ia_coordinates_of_roi',\n",
       " '60': 'CONVERTER_fastqgz_to_fastq',\n",
       " '61': 'mothur_get_communitytype',\n",
       " '62': 'rsem_prepare_reference',\n",
       " '63': 'bedtools_subtractbed',\n",
       " '64': 'cshl_fastx_clipper',\n",
       " '65': 'circgraph',\n",
       " '66': 'bedtools_bamtobed',\n",
       " '67': 'CONVERTER_fasta_to_tabular',\n",
       " '68': 'deeptools_bam_compare',\n",
       " '69': 'mass_spectrometry_imaging_mzplots',\n",
       " '70': 'hicexplorer_hicsummatrices',\n",
       " '71': 'aragorn_trna',\n",
       " '72': 'IDMassAccuracy',\n",
       " '73': 'fastq_quality_trimmer',\n",
       " '74': 'mimodd_varcall',\n",
       " '75': 'mothur_filter_seqs',\n",
       " '76': 'idpqonvert',\n",
       " '77': 'fasta_merge_files_and_filter_unique_sequences',\n",
       " '78': 'Fetch Taxonomic Ranks',\n",
       " '79': 'sklearn_feature_selection',\n",
       " '80': 'qiime_split_libraries_fastq',\n",
       " '81': 'mothur_metastats',\n",
       " '82': 'umi_tools_dedup',\n",
       " '83': 'mothur_dist_seqs',\n",
       " '84': 'rseqc_read_distribution',\n",
       " '85': 'tp_tail_tool',\n",
       " '86': 'edger',\n",
       " '87': 'stacks_procrad',\n",
       " '88': 'snpfinder',\n",
       " '89': 'megablast_wrapper',\n",
       " '90': 'bedtools_slopbed',\n",
       " '91': 'gemini_comp_hets',\n",
       " '92': 'meme_chip',\n",
       " '93': 'sailfish',\n",
       " '94': 'gatk_unified_genotyper',\n",
       " '95': 'mimodd_varextract',\n",
       " '96': 'mothur_rarefaction_single',\n",
       " '97': 'FeatureFinderCentroided',\n",
       " '98': 'minfi_mset',\n",
       " '99': 'mothur_classify_rf',\n",
       " '100': 'sam_to_bam',\n",
       " '101': 'bedtools_mergebed',\n",
       " '102': 'CONVERTER_sam_to_unsorted_bam',\n",
       " '103': 'hmmer_hmmbuild',\n",
       " '104': 'charts',\n",
       " '105': 'MzTabExporter',\n",
       " '106': 'mothur_classify_otu',\n",
       " '107': 'prokka',\n",
       " '108': 'search_gui',\n",
       " '109': 'picard_FilterSamReads',\n",
       " '110': 'Filter1',\n",
       " '111': 'deeptools_plot_fingerprint',\n",
       " '112': 'tabular_to_fastq',\n",
       " '113': 'random_lines1',\n",
       " '114': 'gff2bed1',\n",
       " '115': 'fastq_trimmer',\n",
       " '116': 'CONVERTER_bedgraph_to_bigwig',\n",
       " '117': 'umi_tools_group',\n",
       " '118': 'mimodd_varreport',\n",
       " '119': 'meme_psp_gen',\n",
       " '120': 'samtools_bedcov',\n",
       " '121': 'resistome_analyzer',\n",
       " '122': 'Draw_phylogram',\n",
       " '123': 'tp_find_and_replace',\n",
       " '124': 'unicycler',\n",
       " '125': 'metaphlan_hclust_heatmap',\n",
       " '126': 'PeptideIndexer',\n",
       " '127': 'Multivariate',\n",
       " '128': 'mothur_chimera_uchime',\n",
       " '129': 'gemini_recessive_and_dominant',\n",
       " '130': 'maldi_quant_peak_detection',\n",
       " '131': 'get_feature_info',\n",
       " '132': 'Show beginning1',\n",
       " '133': 'tab2fasta',\n",
       " '134': 'multiqc',\n",
       " '135': 'Count1',\n",
       " '136': 'goseq',\n",
       " '137': 'bwameth',\n",
       " '138': 'je_demultiplex',\n",
       " '139': 'EMBOSS: revseq82',\n",
       " '140': 'IDFilter',\n",
       " '141': 'methtools_destrand',\n",
       " '142': 'mass_spectrometry_imaging_combine',\n",
       " '143': 'cast',\n",
       " '144': 'FROGS_filters',\n",
       " '145': 'iuc_pear',\n",
       " '146': 'idpassemble',\n",
       " '147': 'collection_column_join',\n",
       " '148': 'sklearn_searchcv',\n",
       " '149': 'idr-embl',\n",
       " '150': 'mass_spectrometry_imaging_filtering',\n",
       " '151': 'deeptools_heatmapper',\n",
       " '152': 'cshl_find_and_replace',\n",
       " '153': 'mass_spectrometry_imaging_segmentations',\n",
       " '154': 'mothur_parsimony',\n",
       " '155': 'convert_bc_to_binary_RY.py',\n",
       " '156': 'cshl_fastx_reverse_complement',\n",
       " '157': 'fastq_to_tabular',\n",
       " '158': 'bwa',\n",
       " '159': 'gatk2_variant_recalibrator',\n",
       " '160': 'samtools_stats',\n",
       " '161': 'tp_sort_rows',\n",
       " '162': 'fastq_paired_end_splitter',\n",
       " '163': 'megahit',\n",
       " '164': 'msgfplus',\n",
       " '165': 'table_annovar',\n",
       " '166': 'gatk2_haplotype_caller',\n",
       " '167': 'deeptools_correct_gc_bias',\n",
       " '168': 'heinz_bum',\n",
       " '169': 'picard_CleanSam',\n",
       " '170': 'infernal_cmbuild',\n",
       " '171': 'barchart_gnuplot',\n",
       " '172': 'length_and_gc_content',\n",
       " '173': 'cd_hit_dup',\n",
       " '174': 'dbbuilder',\n",
       " '175': 'signalp3',\n",
       " '176': 'bedtools_intersectbed',\n",
       " '177': 'pileometh',\n",
       " '178': 'gatk2_print_reads',\n",
       " '179': 'peakachu',\n",
       " '180': 'DatamashOps',\n",
       " '181': 'picard_SortSam',\n",
       " '182': 'sickle',\n",
       " '183': 'cufflinks_prok',\n",
       " '184': 'cufflinks',\n",
       " '185': 'bamCompare_deepTools',\n",
       " '186': 'umi_tools_extract',\n",
       " '187': 'taxonomy_krona_chart',\n",
       " '188': 'qiime_validate_mapping_file',\n",
       " '189': 'blast_parser',\n",
       " '190': 'qiime_make_otu_heatmap',\n",
       " '191': 'IsobaricAnalyzer',\n",
       " '192': 'rmcontamination',\n",
       " '193': 'kraken',\n",
       " '194': 'quast',\n",
       " '195': 'merge_pcr_duplicates.py',\n",
       " '196': 'bctools_extract_crosslinked_nucleotides',\n",
       " '197': 'sqlite_to_tabular',\n",
       " '198': 'CometAdapter',\n",
       " '199': 'fasta_filter_by_length',\n",
       " '200': 'hicexplorer_hicfindtads',\n",
       " '201': 'fraggenescan',\n",
       " '202': 'minfi_dmr',\n",
       " '203': 'deeptools_plot_coverage',\n",
       " '204': 'fastq_groomer',\n",
       " '205': 'Grouping1',\n",
       " '206': 'heinz_visualization',\n",
       " '207': 'bamCoverage_deepTools',\n",
       " '208': '__MERGE_COLLECTION__',\n",
       " '209': 'Show tail1',\n",
       " '210': 'peakcalling_macs14',\n",
       " '211': 'gene2exon1',\n",
       " '212': 'bwa_mem',\n",
       " '213': 'scale_image',\n",
       " '214': 'picard_ARRG',\n",
       " '215': 'vcf_filter',\n",
       " '216': 'tp_grep_tool',\n",
       " '217': 'EMBOSS: tranalign100',\n",
       " '218': 'get_sequences',\n",
       " '219': 'tmhmm2',\n",
       " '220': 'vcf2tsv',\n",
       " '221': 'gops_coverage_1',\n",
       " '222': 'dexseq',\n",
       " '223': 'preMloc',\n",
       " '224': 'mothur_chimera_vsearch',\n",
       " '225': 'umi_tools_whitelist',\n",
       " '226': 'ngsutils_bam_filter',\n",
       " '227': 'mothur_nmds',\n",
       " '228': 'hisat2',\n",
       " '229': 'rbc_mafft',\n",
       " '230': 'FeatureLinkerUnlabeledKD',\n",
       " '231': 'gops_intersect_1',\n",
       " '232': 'fastq_paired_end_deinterlacer',\n",
       " '233': 'tophat',\n",
       " '234': 'deseq2',\n",
       " '235': 'qiime_align_seqs',\n",
       " '236': 'viennarna_rnafold',\n",
       " '237': 'EMBOSS: fuzzpro38',\n",
       " '238': 'deeptools_multi_bam_summary',\n",
       " '239': 'cummerbund_to_cuffdiff',\n",
       " '240': 'cummeRbund',\n",
       " '241': 'Univariate',\n",
       " '242': 'minfi_methcpg',\n",
       " '243': 'glimmer_build-icm',\n",
       " '244': 'bedtools_genomecoveragebed_bedgraph',\n",
       " '245': '__UNZIP_COLLECTION__',\n",
       " '246': 'treebest_best',\n",
       " '247': 'proteomics_moff',\n",
       " '248': 'bigwig_to_bedgraph',\n",
       " '249': 'heinz_scoring',\n",
       " '250': 'mothur_remove_groups',\n",
       " '251': 'segemehl',\n",
       " '252': 'cutadapt',\n",
       " '253': 'bam_to_sam',\n",
       " '254': 'trim_reads',\n",
       " '255': 'hicexplorer_hictransform',\n",
       " '256': 'cardinal_preprocessing',\n",
       " '257': 'mothur_cluster_split',\n",
       " '258': 'cshl_fasta_nucleotides_changer',\n",
       " '259': 'PeakPickerHiRes',\n",
       " '260': 'gffcompare',\n",
       " '261': 'CONVERTER_interval_to_bedstrict_0',\n",
       " '262': 'addName',\n",
       " '263': 'p_clip_peaks',\n",
       " '264': 'xpath',\n",
       " '265': 'gops_join_1',\n",
       " '266': 'minfi_getM',\n",
       " '267': 'deeptools_computeMatrix',\n",
       " '268': 'tp_split_on_column',\n",
       " '269': 'picard_AddOrReplaceReadGroups',\n",
       " '270': 'sklearn_build_pipeline',\n",
       " '271': 'DecoyDatabase',\n",
       " '272': 'velvetg_jgi',\n",
       " '273': 'rseqc_geneBody_coverage',\n",
       " '274': 'kallisto_quant',\n",
       " '275': 'gffread',\n",
       " '276': 'Extract_RNA_seq_Evidence',\n",
       " '277': 'biosigner',\n",
       " '278': 'bg_uniq',\n",
       " '279': 'gatk_indel_realigner',\n",
       " '280': 'Paste1',\n",
       " '281': 'mothur_unique_seqs',\n",
       " '282': 'r_correlation_matrix',\n",
       " '283': 'graphlan_annotate',\n",
       " '284': 'humann2_regroup_table',\n",
       " '285': 'mothur_get_seqs',\n",
       " '286': 'flexbar_no_split',\n",
       " '287': 'term_id_vs_term_name',\n",
       " '288': 'deeptools_compute_gc_bias',\n",
       " '289': 'bedtools_intersectbed_bam',\n",
       " '290': '__RELABEL_FROM_FILE__',\n",
       " '291': 'mothur_summary_single',\n",
       " '292': 'qiime_core_diversity',\n",
       " '293': 'vegan_diversity',\n",
       " '294': 'abims_xcms_group',\n",
       " '295': 'mothur_remove_lineage',\n",
       " '296': 'vcfallelicprimitives',\n",
       " '297': 'cshl_uniq_tool',\n",
       " '298': 'gatk2_unified_genotyper',\n",
       " '299': 'cuffquant',\n",
       " '300': 'gemini_query',\n",
       " '301': 'mothur_taxonomy_to_krona',\n",
       " '302': 'humann2_genefamilies_genus_level',\n",
       " '303': 'genbank_to_gff',\n",
       " '304': 'featurecounts',\n",
       " '305': 'viennarna_rnaplot',\n",
       " '306': 'format_metaphlan2_output',\n",
       " '307': 'cshl_awk_tool',\n",
       " '308': 'nspdk_sparse',\n",
       " '309': 'bedtools_sortbed',\n",
       " '310': 'gtftobed12',\n",
       " '311': 'naive_variant_caller',\n",
       " '312': 'samtools_filter',\n",
       " '313': 'cardinal_segmentations',\n",
       " '314': 'cshl_fastq_to_fasta',\n",
       " '315': 'bed2gff1',\n",
       " '316': 'valet',\n",
       " '317': 'methtools_calling',\n",
       " '318': 'ggplot2_histogram',\n",
       " '319': 'ip_projective_transformation_points',\n",
       " '320': 'filtlong',\n",
       " '321': 'bedtools_coveragebed',\n",
       " '322': 'picard_CASM',\n",
       " '323': 'annotateMyIDs',\n",
       " '324': 'samtools_slice_bam',\n",
       " '325': 'FROGS_remove_chimera',\n",
       " '326': 'cardinal_data_exporter',\n",
       " '327': 'ete_genetree_splitter',\n",
       " '328': 'minfi_qc',\n",
       " '329': 'MassCalculator',\n",
       " '330': 'get_subontology_from',\n",
       " '331': 'comp1',\n",
       " '332': 'locarna_best_subtree',\n",
       " '333': 'CONVERTER_bed_to_gff_0',\n",
       " '334': 'replace_column_with_key_value_file',\n",
       " '335': 'macs2_callpeak',\n",
       " '336': 'graphlan',\n",
       " '337': 'igvtools_count',\n",
       " '338': 'jbrowse',\n",
       " '339': 'FROGS_affiliation_OTU',\n",
       " '340': 'bwa_wrapper',\n",
       " '341': 'rseqc_bam2wig',\n",
       " '342': 'umi_tools_count',\n",
       " '343': 'venn_list',\n",
       " '344': 'gemini_de_novo',\n",
       " '345': 'picard_MarkDuplicates',\n",
       " '346': 'heinz',\n",
       " '347': 'annotatemyids',\n",
       " '348': 'coords2clnt.py',\n",
       " '349': 'samtools_flagstat',\n",
       " '350': 'samtools_sort',\n",
       " '351': 'mimodd_map',\n",
       " '352': 'rbc_mirdeep2_mapper',\n",
       " '353': 'modencode_peakcalling_spp',\n",
       " '354': 'deeptools_plot_pca',\n",
       " '355': 'vcftools_merge',\n",
       " '356': 'piranha',\n",
       " '357': 'tp_multijoin_tool',\n",
       " '358': 'samtools_idxstats',\n",
       " '359': 'minfi_dropsnp',\n",
       " '360': 'MapRTTransformer',\n",
       " '361': 'idpquery',\n",
       " '362': 'samtools_rmdup',\n",
       " '363': 'tp_replace_in_line',\n",
       " '364': 'mergeCols1',\n",
       " '365': 'mothur_cluster',\n",
       " '366': 'gspan',\n",
       " '367': 'GeneModelMapper',\n",
       " '368': 'picard_ReorderSam',\n",
       " '369': 'IDMerger',\n",
       " '370': 'mothur_homova',\n",
       " '371': 'rseqc_bam_stat',\n",
       " '372': 'mothur_pcoa',\n",
       " '373': 'mothur_heatmap_bin',\n",
       " '374': 'tp_awk_tool',\n",
       " '375': 'bowtie_wrapper',\n",
       " '376': 'ete_homology_classifier',\n",
       " '377': 'FalseDiscoveryRate',\n",
       " '378': 'hgv_david',\n",
       " '379': 'bedtools_windowbed',\n",
       " '380': 'pureclip',\n",
       " '381': 'sed_stream_editor',\n",
       " '382': 'viennarna_rnaalifold',\n",
       " '383': 'samtool_filter2',\n",
       " '384': 'pieplots_macs',\n",
       " '385': 'mlocarna',\n",
       " '386': 'fastq_to_fasta_python',\n",
       " '387': 'mothur_pre_cluster',\n",
       " '388': 'collapse_dataset',\n",
       " '389': 'Remove beginning1',\n",
       " '390': 'ChangeCase',\n",
       " '391': 'mothur_heatmap_sim',\n",
       " '392': 'mothur_seq_error',\n",
       " '393': 'allele_counts_1',\n",
       " '394': 'fastq_filter',\n",
       " '395': 'IDMapper',\n",
       " '396': 'kraken-report',\n",
       " '397': 'hcluster_sg_parser',\n",
       " '398': 'bam_to_bigwig',\n",
       " '399': 'regex_replace',\n",
       " '400': 'meme_fimo',\n",
       " '401': 'velvet',\n",
       " '402': 'Grep1',\n",
       " '403': 'FidoAdapter',\n",
       " '404': 'tp_sort_header_tool',\n",
       " '405': 'bedtools_genomecoveragebed',\n",
       " '406': 'EMBOSS: transeq101',\n",
       " '407': 'CONVERTER_gz_to_uncompressed',\n",
       " '408': 'deeptools_plot_enrichment',\n",
       " '409': 'vcfbedintersect',\n",
       " '410': 'XTandemAdapter',\n",
       " '411': 'abims_xcms_fillPeaks',\n",
       " '412': 'tp_replace_in_column',\n",
       " '413': 'get_flanks1',\n",
       " '414': 'ncbi_blastp_wrapper',\n",
       " '415': 'gatk2_realigner_target_creator',\n",
       " '416': 'blastxml_to_top_descr',\n",
       " '417': 'gops_merge_1',\n",
       " '418': 'metaphlan2krona',\n",
       " '419': 'column_remove_by_header',\n",
       " '420': 'freebayes',\n",
       " '421': 'stacks_populations',\n",
       " '422': 'sklearn_nn_classifier',\n",
       " '423': 'qiime_count_seqs',\n",
       " '424': 'cardinal_quality_report',\n",
       " '425': 'fastq_combiner',\n",
       " '426': 'meme_dreme',\n",
       " '427': 'bowtie2',\n",
       " '428': 'EMBOSS: getorf42',\n",
       " '429': 'wiggle2simple1',\n",
       " '430': 'preproc',\n",
       " '431': 'XY_Plot_1',\n",
       " '432': 'graphicsmagick_image_convert',\n",
       " '433': 'mimodd_vcf_filter',\n",
       " '434': 'picard_CollectRnaSeqMetrics',\n",
       " '435': 'FeatureLinkerUnlabeledQT',\n",
       " '436': 'fasta_compute_length',\n",
       " '437': 'viz_overlay_moving_and_fixed_image',\n",
       " '438': 'nn_classifier',\n",
       " '439': 'bedtools_getfastabed',\n",
       " '440': 'ggplot2_heatmap2',\n",
       " '441': 'cshl_fastx_renamer',\n",
       " '442': 'mothur_make_shared',\n",
       " '443': 'addValue',\n",
       " '444': 'tophat2',\n",
       " '445': 'tp_text_file_with_recurring_lines',\n",
       " '446': 'abims_xcms_xcmsSet',\n",
       " '447': '__FLATTEN__',\n",
       " '448': 'tp_sorted_uniq',\n",
       " '449': 'mothur_count_groups',\n",
       " '450': 'seq_filter_by_id',\n",
       " '451': 'mothur_merge_files',\n",
       " '452': 'rgPicardMarkDups',\n",
       " '453': 'gatk2_reduce_reads',\n",
       " '454': 'stringtie',\n",
       " '455': 'hmmer_hmmsearch',\n",
       " '456': 'bedtools_map',\n",
       " '457': 'bamleftalign',\n",
       " '458': 'cardinal_filtering',\n",
       " '459': 'snpSift_dbnsfp_generic',\n",
       " '460': 'peptide_shaker',\n",
       " '461': 'flexbardsc',\n",
       " '462': 'IDConflictResolver',\n",
       " '463': 'stacks_denovomap',\n",
       " '464': 'cuffcompare',\n",
       " '465': 'racon',\n",
       " '466': 'gops_concat_1',\n",
       " '467': 'mass_spectrometry_imaging_qc',\n",
       " '468': 'samtools_mpileup',\n",
       " '469': 'cuffdiff',\n",
       " '470': 'trinity',\n",
       " '471': 'mothur_dist_shared',\n",
       " '472': 'mothur_remove_seqs',\n",
       " '473': 'minfi_read450k',\n",
       " '474': 'cshl_fasta_formatter',\n",
       " '475': 'plotly_parallel_coordinates_plot',\n",
       " '476': 'extract_bcs.py',\n",
       " '477': 'IDPosteriorErrorProbability',\n",
       " '478': 'NFS_transfer',\n",
       " '479': 'gff_filter_by_attribute',\n",
       " '480': 'snpSift_filter',\n",
       " '481': 'ncbi_tblastn_wrapper',\n",
       " '482': 'blockclust',\n",
       " '483': 'wc_gnu',\n",
       " '484': 'humann2',\n",
       " '485': 'tophat_fusion_post',\n",
       " '486': 'fastqc',\n",
       " '487': 'Heatmap',\n",
       " '488': 'glob_report',\n",
       " '489': 'Extract genomic DNA 1',\n",
       " '490': 'hicexplorer_hiccorrectmatrix',\n",
       " '491': 'tp_cat',\n",
       " '492': 'snpEff_download',\n",
       " '493': 'deeptools_bigwig_compare',\n",
       " '494': 'computeMatrix',\n",
       " '495': 'bamFilter',\n",
       " '496': 'deeptools_plot_heatmap',\n",
       " '497': 'spades',\n",
       " '498': 'extract_aln_ends.py',\n",
       " '499': 'hicexplorer_hicbuildmatrix',\n",
       " '500': 'tp_sed_tool',\n",
       " '501': 'cshl_fastx_collapser',\n",
       " '502': 'picard_SamToFastq',\n",
       " '503': 'deeptools_bamCoverage',\n",
       " '504': 'bundle_collection',\n",
       " '505': 'wsdl_hmdb',\n",
       " '506': 'velveth_jgi',\n",
       " '507': 'group_humann2_uniref_abundances_to_go',\n",
       " '508': 'minfi_ppfun',\n",
       " '509': 'TextExporter',\n",
       " '510': 'sort1',\n",
       " '511': 'MAF_To_Fasta1',\n",
       " '512': 'deeptools_compute_matrix_operations',\n",
       " '513': 'ProteinResolver',\n",
       " '514': 'mothur_get_oturep',\n",
       " '515': 'hicexplorer_hicmergematrixbins',\n",
       " '516': 'ncbi_blastn_wrapper',\n",
       " '517': 'picard_MarkDuplicatesWithMateCigar',\n",
       " '518': 'bg_find_subsequences',\n",
       " '519': 'megablast_xml_parser',\n",
       " '520': 'mothur_align_seqs',\n",
       " '521': 'gatk2_indel_realigner',\n",
       " '522': 'je_clip',\n",
       " '523': 'query_tabular',\n",
       " '524': 'xcms_merge',\n",
       " '525': 'gafa',\n",
       " '526': '__SORTLIST__',\n",
       " '527': 'hgv_snpFreq',\n",
       " '528': 'picard_FixMateInformation',\n",
       " '529': 'FROGS_preprocess',\n",
       " '530': 'fastq_paired_end_joiner',\n",
       " '531': 'deeptools_compute_matrix',\n",
       " '532': 'meme_meme',\n",
       " '533': 'rsem_calculate_expression',\n",
       " '534': 'bcftools_call',\n",
       " '535': 'lfq_protein_quant',\n",
       " '536': 'dt_profiler',\n",
       " '537': 'cshl_fastx_barcode_splitter',\n",
       " '538': 'bam-filter',\n",
       " '539': 'trim_galore',\n",
       " '540': 'combine_metaphlan2_humann2',\n",
       " '541': 'rcas',\n",
       " '542': 'filter_by_fasta_ids',\n",
       " '543': 'rarefaction_analyzer',\n",
       " '544': 'varscan_copynumber',\n",
       " '545': 'cshl_word_list_grep',\n",
       " '546': 'column_order_header_sort',\n",
       " '547': 'fastq_join',\n",
       " '548': 'w4mclstrpeakpics',\n",
       " '549': 'cshl_sort_header',\n",
       " '550': 'bctools_remove_spurious_events',\n",
       " '551': 'vcfsort',\n",
       " '552': 'blockbuster',\n",
       " '553': 'FROGS_clustering',\n",
       " '554': 'snpSift_annotate',\n",
       " '555': 'w4mjoinpn',\n",
       " '556': 'lca1',\n",
       " '557': 'fasta-stats',\n",
       " '558': 'ProteinQuantifier',\n",
       " '559': 'Convert characters1',\n",
       " '560': 'remove_tail.py',\n",
       " '561': 'cat1',\n",
       " '562': 'tp_easyjoin_tool',\n",
       " '563': 'hcluster_sg',\n",
       " '564': 'bg_protein_properties',\n",
       " '565': 'tsne',\n",
       " '566': 'join1',\n",
       " '567': 'mothur_sub_sample',\n",
       " '568': 'cshl_cut_tool',\n",
       " '569': 'varscan_somatic',\n",
       " '570': 'augustus',\n",
       " '571': 'mothur_venn',\n",
       " '572': 'fasta2tab',\n",
       " '573': 'dexseq_annotate',\n",
       " '574': 'bctools_remove_tail',\n",
       " '575': 'cuffmerge',\n",
       " '576': 'myrimatch',\n",
       " '577': 'FileMerger',\n",
       " '578': 'nanopolish_variants',\n",
       " '579': 'cshl_fastx_trimmer',\n",
       " '580': 'rbc_mirdeep2',\n",
       " '581': 'Summary_Statistics1',\n",
       " '582': 'trimmer',\n",
       " '583': 're_he_maldi_image_registration',\n",
       " '584': 'ProteinInference',\n",
       " '585': 'HighResPrecursorMassCorrector',\n",
       " '586': 'w4mclassfilter',\n",
       " '587': 'seq_composition',\n",
       " '588': 'rna_star',\n",
       " '589': 'hicexplorer_hicpca',\n",
       " '590': 'stacks_genotypes',\n",
       " '591': 'vegan_rarefaction',\n",
       " '592': 'sample_seqs',\n",
       " '593': 'MSGFPlusAdapter',\n",
       " '594': '__FILTER_FROM_FILE__',\n",
       " '595': 'mothur_make_group',\n",
       " '596': '__FILTER_FAILED_DATASETS__',\n",
       " '597': 'FileFilter',\n",
       " '598': 'qiime_pick_open_reference_otus',\n",
       " '599': 'sam_merge2',\n",
       " '600': 'mass_spectrometry_imaging_classification',\n",
       " '601': 'MapStatistics',\n",
       " '602': 'trimmomatic',\n",
       " '603': 't2t_report',\n",
       " '604': 'mothur_screen_seqs',\n",
       " '605': 'diffbind-embl',\n",
       " '606': 'cshl_fastq_quality_filter',\n",
       " '607': 'metaphlan2',\n",
       " '608': 'melt',\n",
       " '609': 'mass_spectrometry_imaging_ion_images',\n",
       " '610': 'stacks_refmap',\n",
       " '611': 'fasplit',\n",
       " '612': 'roary',\n",
       " '613': 'cmFinder',\n",
       " '614': 'bctools_extract_alignment_ends',\n",
       " '615': 'MapAlignerPoseClustering',\n",
       " '616': 'antismash',\n",
       " '617': 'gemini_actionable_mutations',\n",
       " '618': 'maldi_quant_preprocessing',\n",
       " '619': 'cor2',\n",
       " '620': 'varscan',\n",
       " '621': 'minfi_getbeta',\n",
       " '622': 'fasttree',\n",
       " '623': 'mothur_get_groups',\n",
       " '624': 'rseqc_inner_distance',\n",
       " '625': 'gemini_load',\n",
       " '626': 'minimap2',\n",
       " '627': 'FileInfo',\n",
       " '628': 'deeptools_bam_coverage',\n",
       " '629': 'stringtie_merge',\n",
       " '630': 'disco',\n",
       " '631': 'CONVERTER_interval_to_bed_0',\n",
       " '632': 'CONVERTER_gff_to_bed_0',\n",
       " '633': 'export2graphlan',\n",
       " '634': 'heatmapper_deepTools',\n",
       " '635': 'vcffilter2',\n",
       " '636': 'mothur_make_contigs',\n",
       " '637': 't_coffee',\n",
       " '638': 'snippy',\n",
       " '639': 'flye',\n",
       " '640': 'FileConverter',\n",
       " '641': 'vcftools_annotate',\n",
       " '642': 'minfi_getanno',\n",
       " '643': 'metaspades',\n",
       " '644': 'cardinal_mz_images',\n",
       " '645': 'bed_to_bigwig',\n",
       " '646': 'infernal_cmsearch',\n",
       " '647': 'MultiplexResolver',\n",
       " '648': 'methtools_plot',\n",
       " '649': 'structure_to_gspan',\n",
       " '650': 'join_files_on_column_fuzzy',\n",
       " '651': 'SequenceCoverageCalculator',\n",
       " '652': 'QCCalculator',\n",
       " '653': 'datamash_ops',\n",
       " '654': 'deeptools_plot_profile',\n",
       " '655': 'deeptools_multi_bigwig_summary',\n",
       " '656': 'htseq_count',\n",
       " '657': 'hicexplorer_hicplottads',\n",
       " '658': '__APPLY_RULES__',\n",
       " '659': 'gops_subtract_1',\n",
       " '660': 'histogram_rpy',\n",
       " '661': 'je_markdupes',\n",
       " '662': 'vsearch_dereplication',\n",
       " '663': 'CONVERTER_fasta_to_fai',\n",
       " '664': 'IDFileConverter',\n",
       " '665': 'mothur_phylotype',\n",
       " '666': 'mothur_corr_axes',\n",
       " '667': 'samtools_flag_filter',\n",
       " '668': 'ncbi_makeblastdb',\n",
       " '669': 'hicexplorer_hicplotmatrix',\n",
       " '670': 'decoyfasta',\n",
       " '671': 'GeMoMa_Annotation_Filter',\n",
       " '672': 'rm_spurious_events.py',\n",
       " '673': 'FeatureFinderMultiplex',\n",
       " '674': 'snpEff',\n",
       " '675': 'mothur_make_biom',\n",
       " '676': 'rseqc_infer_experiment',\n",
       " '677': 'nonpareil',\n",
       " '678': 'bg_sortmerna',\n",
       " '679': 'fastq_dump',\n",
       " '680': 'salmon',\n",
       " '681': 'vcfcombine',\n",
       " '682': 'checkFormat',\n",
       " '683': 'gatk_realigner_target_creator',\n",
       " '684': 'thermo_raw_file_converter',\n",
       " '685': 'abims_xcms_summary',\n",
       " '686': 'tp_cut_tool',\n",
       " '687': 'glimmer_knowlegde-based',\n",
       " '688': 'Interval_Maf_Merged_Fasta2',\n",
       " '689': 'Extractor',\n",
       " '690': 'humann2_renorm_table',\n",
       " '691': 'rgPicFixMate',\n",
       " '692': 'fastq_manipulation',\n",
       " '693': 'ConsensusMapNormalizer'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reverse_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tool seq: trimmomatic,bwa_mem\n",
      "\n",
      "Predicted tools: \n",
      "\n",
      "{'samtools_sort': 0.99, 'bam_to_sam': 0.99, 'samtools_flagstat': 0.99, 'hicexplorer_hicbuildmatrix': 0.98, 'samtools_stats': 0.97, 'samtools_idxstats': 0.96, 'samtools_filter': 0.96, 'samtool_filter2': 0.94, 'picard_MarkDuplicates': 0.94, 'samtools_rmdup': 0.92, 'samtools_mpileup': 0.81, 'deeptools_bam_coverage': 0.81, 'ngsutils_bam_filter': 0.71, 'rseqc_bam2wig': 0.64, 'picard_AddOrReplaceReadGroups': 0.64}\n",
      "\n",
      "Class weights: \n",
      "\n",
      "{'deeptools_bam_coverage': 6.88, 'hicexplorer_hicbuildmatrix': 6.33, 'picard_MarkDuplicates': 5.53, 'samtool_filter2': 5.39, 'picard_AddOrReplaceReadGroups': 5.3, 'samtools_idxstats': 5.22, 'samtools_rmdup': 5.08, 'samtools_flagstat': 4.9, 'samtools_sort': 4.44, 'samtools_mpileup': 4.39, 'bam_to_sam': 4.0, 'samtools_filter': 3.8, 'rseqc_bam2wig': 3.74, 'ngsutils_bam_filter': 3.65, 'samtools_stats': 3.54}\n",
      "\n",
      "Usage weights: \n",
      "\n",
      "{'deeptools_bam_coverage': 1243.06, 'picard_MarkDuplicates': 352.98, 'samtools_flagstat': 274.22, 'samtools_rmdup': 253.46, 'samtool_filter2': 235.75, 'samtools_idxstats': 208.46, 'hicexplorer_hicbuildmatrix': 115.62, 'samtools_mpileup': 108.15, 'samtools_sort': 106.9, 'ngsutils_bam_filter': 80.1, 'bam_to_sam': 70.28, 'samtools_stats': 57.96, 'rseqc_bam2wig': 19.54, 'samtools_filter': 15.8, 'picard_AddOrReplaceReadGroups': 10.3}\n",
      "\n",
      "Mean usage wt: 210.1720\n",
      "\n",
      "Inverted weights: \n",
      "\n",
      "{}\n",
      "\n",
      "Tool ids\n",
      "{'samtools_sort': 350, 'bam_to_sam': 253, 'samtools_flagstat': 349, 'hicexplorer_hicbuildmatrix': 499, 'samtools_stats': 160, 'samtools_idxstats': 358, 'samtools_filter': 312, 'samtool_filter2': 383, 'picard_MarkDuplicates': 345, 'samtools_rmdup': 362, 'samtools_mpileup': 468, 'deeptools_bam_coverage': 628, 'ngsutils_bam_filter': 226, 'rseqc_bam2wig': 341, 'picard_AddOrReplaceReadGroups': 269}\n",
      "======================================\n"
     ]
    }
   ],
   "source": [
    "topk = 20\n",
    "tool_seq = \"602,212\"\n",
    "class_wt, usage_wt, inverse_wt, pred_tools = verify_model(model, tool_seq, \"\", dictionary, reverse_dictionary, compatibile_tools, 1, topk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
